src/ralph/converters/xapi/base.py

```diff
@@ -0,0 +1,126 @@
+    @post_dump
+    def get_xapi_statement(self, data, many):  # pylint: disable=unused-argument
+        """Returns the xApi Statement"""
+        return json.loads(Statement(**self.xapi_props).to_json())
+
+    def add_required_properties(self):
+        """Add the required actor, verb, object properies to xapi_props"""
+        self.xapi_props["actor"] = self.get_actor()
+        self.xapi_props["verb"] = Verb(**self.get_verb_kwargs())
+        self.xapi_props["object"] = self.get_object()
+
+    def add_optional_properties(self):
+        """Add the optional result, context, timestamp properies to xapi_props"""
+        self.add_optional_property("result", self.get_result())
+        self.add_optional_property("context", self.get_context())
+        self.add_optional_property("timestamp", self.get_timestamp())
+
+    def add_optional_property(self, name, value):
+        """Add an optional property by given name if the value is not None"""
+        if value is not None:
+            self.xapi_props[name] = value
+
+    def get_actor(self):
+        """Returns the actor property to xapi_props"""
+        name = self.edx_event["username"]
+        if not name:
+            name = "anonymous"
+        account = AgentAccount(name=name, home_page=HOME_PAGE)
+        return Agent(account=account)
+
+    def get_verb_kwargs(self):
+        """Returns tincan.Verb kwargs as a dict for xapi_props
+        should contain id and display key value pairs
+        """
+        raise NotImplementedError("get_verb_kwargs should be implemented")
+
+    def get_object(self):
+        """Returns the object property for xapi_props
+        it might be a tincan Activity, Agent, SubStatement or StatementRef
+        """
+        raise NotImplementedError("get_object should be implemented")
+
+    def get_result(self):  # pylint: disable=no-self-use
+        """Returns the result property for xapi_props"""
+        return None
+
+    def get_context(self):
+        """Returns the context property for xapi_props"""
+        return Context(
+            extensions=self.get_common_context_extensions(), platform=HOME_PAGE
+        )
+
+    def get_common_context_extensions(self):
+        """Returns the extensions property for xapi_props context"""
+        context = self.edx_event["context"]
+        event_ip = self.edx_event["ip"]
+        if event_ip:
+            event_ip = event_ip.exploded
+        extensions = {
+            XAPI_EXTENSION_ACCEPT_LANGUAGE: self.edx_event["accept_language"],
+            XAPI_EXTENSION_AGENT: self.edx_event["agent"],
+            XAPI_EXTENSION_COURSE_ID: context["course_id"],
+            XAPI_EXTENSION_COURSE_USER_TAGS: context.get("course_user_tags", {}),
+            XAPI_EXTENSION_HOST: self.edx_event["host"],
+            XAPI_EXTENSION_IP: event_ip,
+            XAPI_EXTENSION_ORG_ID: context["org_id"],
+            XAPI_EXTENSION_REFERER: self.edx_event["referer"],
+            XAPI_EXTENSION_USER_ID: context["user_id"],
```

Comment:
XAPI_EXTENSION_USER_ID can be null and null values are removed from xApi statements.
Do you prefer to have an empty XAPI_EXTENSION_USER_ID value instead or we can keep this behaviour?

Reply:
The current behaviour is: we keep the XAPI_EXTENSION_USER_ID and set it to an empty string when user_id is null